### ISSUE TITLE
glob false positives

### ISSUE THREAD

**Issue by dereuromark** [vimeo/psalm#3796]
Our code has a few like these
```php
    return array_filter($lookupPaths, 'glob');
```
which is valid and working PHP code afaik. Using functions like this.

Psalm reports errors like
```json
            "type": "InvalidArgument",
            "message": "Argument 2 of array_filter expects callable(mixed, mixed=):scalar, string(glob) provided",
            "file_name": ".../PropelSchemaParser.php",
            "snippet": "        return array_filter($lookupPaths, 'glob');",
            "selected_text": "'glob'",
```

For
```php
        $directories = array_filter($directories, function (string $directory) {
            return glob($directory);
        });
```
it reports
```json
            "type": "InvalidArgument",
            "message": "Argument 2 of array_filter expects callable(mixed, mixed=):scalar, Closure(string):(false|list<string>) provided",
            "file_name": ".../FileFinder.php",
            "snippet": "        $directories = array_filter($directories, function (string $directory) {\n            return glob($directory);\n        });",
            "selected_text": "function (string $directory) {\n            return glob($directory);\n        }",
```

**Comment by dereuromark** [comment]
https://psalm.dev/r/75ed5813fb

**Comment by muglug** [comment]
I don't really think this is a false-positive – I think being explicit about how the `glob` call is used makes the code easier to understand:

```php
$result = array_filter($lookupPaths, fn($path) => (bool) glob($path));
```

**Comment by dereuromark** [comment]
return array_filter($lookupPaths, 'glob');

is valid PHP though

**Comment by 9brada6** [comment]
> is valid PHP though

@dereuromark  True, that's the first example on [php docs](https://www.php.net/manual/en/function.array-filter.php), and I also didn't understand why it is false-positive, but after thinking a little bit I understand and I will explain:

>  Argument 2 of array_filter expects callable(mixed, mixed=):**scalar**

Look on the return, it expects to return a scalar. Glob return is not scalar. From PHP DOC:

> Scalar variables are those containing an integer, float, string or boolean. Types array, object and resource are not scalar.

Using a function that returns a scalar it won't report any error:

https://psalm.dev/r/b789bfbe1f

Also, in the second argument you should pass a function that return either true or false(glob return an array), and is explicit from the name to be nicer to read :).
